lib/okta/directory_service.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require 'common/client/base'
+
+module Okta
+  class DirectoryService < Common::Client::Base
+    def get_apps
+      okta_service = Okta::Service.new
+      redis = Redis.new
+      base_url = Settings.oidc.base_api_url + '/api/v1/apps?limit=200&filter=status+eq+\"ACTIVE\"'
+      unfiltered_apps = recursively_get_apps(okta_service, base_url)
+
+      # matches any ISO date.
+      # rubocop:disable Metrics/LineLength
+      iso_pattern = /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)/
+      # rubocop:enable Metrics/LineLength
+
+      # Iterate through the returned applications, and test for pattern matching,
+      # adding to our filtered apps array if pattern doesn't match
+      filtered_apps = unfiltered_apps.reject { |app| app['label'] =~ iso_pattern }
+
+      # This feature is currently in early-access and access will need to be requested from okta.
+      # We are currently using the EA feature to get all our user grants in the connected-applications
+      # profile feature so this shouldn't be an issue to add as well.
+      # get_scopes(apps)
+
+      redis.set('okta_directory_apps', filtered_apps.to_json)
+      unfiltered_apps
+    end
+
+    def recursively_get_apps(okta_service, url = '', unfiltered_apps = [])
+      apps_response = okta_service.get_apps(url)
+      # Moving apps in response body to iterable array
+      unfiltered_apps.concat(apps_response.body)
+
+      # Check headers for ['link'] where 'rel' == next
+      # If the next link exists, call okta_service.get_apps(next_link) and filter based on iso_pattern
+      if contains_next(apps_response.headers)
+        next_link = substring_next_link(apps_response.headers)
+        recursively_get_apps(okta_service, next_link, unfiltered_apps)
+      end
+
+      unfiltered_apps
+    end
+
+    # Check if headers contains 'next' link, since 'next' is always after 'self', we can use .last as a consistent check
+    def contains_next(headers)
+      headers['link'].split(',').last.include?('next')
+    end
+
+    def substring_next_link(headers)
+      headers['link'].split(',').last.split('<').last.split('>').first
```

Comment:
Include some &. to the front of those methods.

Reply:
Done!